module/x/peggy/abci.go

```diff
@@ -0,0 +1,141 @@
+		// and we slash users who haven't signed a valset that is currentHeight - signedBlocksWindow old
+		case uint64(ctx.BlockHeight())-params.SignedValsetsWindow > vs.Height:
+
+			// first we need to see which validators in the active set
+			// haven't signed the valdiator set and slash them,
+			var toSlash []stakingtypes.Validator
+			confirms := k.GetValsetConfirms(ctx, vs.Nonce)
+			for _, val := range currentBondedSet {
+				found := false
+				for _, conf := range confirms {
+					if conf.EthAddress == k.GetEthAddress(ctx, val.GetOperator()) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					toSlash = append(toSlash, val)
+				}
+			}
+
+			for _, val := range toSlash {
+				cons, _ := val.GetConsAddr()
+				k.StakingKeeper.Slash(ctx, cons, ctx.BlockHeight(), val.ConsensusPower(), params.SlashFractionValset)
+				k.StakingKeeper.Jail(ctx, cons)
+			}
+
+			// then we prune the valset from state
+			k.DeleteValset(ctx, vs.Nonce)
+
+		// on the latest validator set, check for change in power against
+		// current, and emit a new validator set if the change in power >1%
+		case i == 0:
+			if types.BridgeValidators(k.GetCurrentValset(ctx).Members).PowerDiff(vs.Members) > 0.01 {
+				k.SetValsetRequest(ctx)
+			}
+		}
+	}
+
+	// #2 condition
+	// We look through the full bonded set (not just the active set, include unbonding validators)
+	// and we slash users who haven't signed a batch confirmation that is >15hrs in blocks old
+	batches := k.GetOutgoingTxBatches(ctx)
+	for _, batch := range batches {
+		if uint64(ctx.BlockHeight())-params.SignedBatchesWindow > batch.Block {
+			var toSlash []stakingtypes.Validator
+			confirms := k.GetBatchConfirmByNonceAndTokenContract(ctx, batch.BatchNonce, batch.TokenContract)
+			for _, val := range currentBondedSet {
+				found := false
+				for _, conf := range confirms {
+					// TODO: may need to look up actual validator address
+					confVal, _ := sdk.AccAddressFromBech32(conf.Orchestrator)
+					if confVal.Equals(val.GetOperator()) {
+						found = true
+					}
+				}
+				if !found {
+					toSlash = append(toSlash, val)
+				}
+			}
+			for _, val := range toSlash {
+				cons, _ := val.GetConsAddr()
+				// TODO: make this a different slash fraction in the params
+				k.StakingKeeper.Slash(ctx, cons, ctx.BlockHeight(), val.ConsensusPower(), params.SlashFractionBatch)
+				k.StakingKeeper.Jail(ctx, cons)
+			}
+
+			// clean up batches here
+			k.DeleteBatch(ctx, *batch)
+		}
+	}
+
+	// #3 condition
+	// Oracle events MsgDepositClaim, MsgWithdrawClaim
+	attmap := k.GetAttestationMapping(ctx)
+	for _, atts := range attmap {
+		// Conflicting votes should be slashed
+		if len(atts) > 1 {
+			var (
+				toSlash []string
+				unObs   []types.Attestation
+			)
+			oneObserved := false
+			for _, att := range atts {
+				if att.Observed == true {
+					oneObserved = true
+					continue
+				}
+				unObs = append(unObs, att)
```

Comment:
check if oneObserved and slash. else add to the list so that you don't have to iterate twice over the items that are after the first observed attestation